### PR TITLE
eth: request id dispatcher and direct req/reply APIs

### PR DESCRIPTION
**Pull request by karalabe**
With all non-request-id `eth` protocol version dropped (`<= eth/65`), we can finally properly match up inbound replies to pending requests (e.g. inbound `ReceiptsMsg` to pending `GetReceiptsMsg`). This is a very good thing, as is greatly simplifies internal code... after a huge rewrite, this PR.

### Background

A huge complexity in our network packet handling code is that there can be concurrent requests of the same type sent to the same peer. These happen because there are multiple subsystems that run concurrently, and which may want to fetch something:

- The `downloader` requests headers, bodies, receipts and states.
- The `fetcher` requests headers and bodies.
- The `eth` package does header challenges.

To make matters worse, if a peer disconnects and quickly reconnects (Parity liked to do this), it can happen that there are still some reply handling propagating through the system from the old connection, whilst the new connection is hit with requests and starts delivering replies too.

The net effect is that when a reply packet arrives (header, body, receipt, state delivery), there can be many requests waiting for it all over the place. Since there's no way in <= eth/65 to match the arriving packets up with the requests, our code had a  filtering mechanism implemented: a reply packet is first given to the eth challenge handler; it it doesn't consume it, it is handed to the fetcher; if it is still not consumed, it is handed to the downloader.

This quickly becomes a horrible thing when we also allow partial deliveries or altogether empty replies (missing data); as subsystems will start to consume each others' data, leading to inconsistencies (timeouts and weird behavior) in the starved subsystem. Further complexity arises when peers send spam data; or when a reply reaches a subsystem where the peer already disconnected/reconnected and was issues a new request.

Geth does handle all this complexity currently. It's just insanely complicated and with The Merge coming and requiring heavy protocol changes; it's extremely brittle to keep hacking the existing code, hoping it won't blow up. The only sane way out, is to rework the entire reply consumption mechanism; or formally put, the reply dispatching.

### Dispatcher

As detailed above, Geth's reply handling is insanely complicated, forced by the `eth` protocol not supporting matching replies to requests. `eth/66`, however, introduced a request ID in all request packets that get echoed back in the replies to those packets.

Using these ping-ponged IDs, we could ensure that subsystems don't consume each other's messages; that unrequested data gets correctly dropped; or that deliveries across reconnects don't influence each other. Still, keeping the old architecture of each subsystem filtering through all the inbound messages - just because in the legacy protocols they had to - becomes useless complexity that still hinders work on The Merge.

Instead of having an architecture where network replies enter the system through the protocol handlers and then bubble through everything to reach their final destination, this PR introduces a request/reply dispatcher that delivers a reply message directly to the point of request. This simplifies packet handling a lot:

- If a subsystem does a single request to a peer and wants to wait for the response (e.g. the `eth` chain challenges; or the `downloader` ancestor lookup); it can simply block - inline in the code, like a function call waiting for the return - until the reply arrives or the request times out. No more waiting on channels that might deliver junk.
- If a subsystem does concurrent requests to many peers (e.g. the `downloader` receipt chain retrieval), it can multiplex the replies into a single sink channel, but still have the guarantee that only one reply arrives to a request and that every request either gets serviced or times out. This moves the error handling complexity out of the higher level subsystem.

This dispatcher is implemented within `eth/protocols/eth` and is part of the `eth.Peer` (i.e. each peer gets its own dispatcher to avoid starvation issues). The dispatcher in essence is a goroutine that receives `eth.Request` objects to send over the wire and tracks these until a response-type message arrives bundled up as an `eth.Response`. After a successful match, the response payload is delivered on the sink channel of the request.

The `eth.Request` objects aren't constructed publicly, they are created by `eth.Peer.RequestXYZ`, but they are returned to the caller since they allow cancelling a pending request if the caller deems so (e.g. timeout, disconnect, etc). It is **essential** to note that the dispatcher **does not** untrack requests on disconnect and **does not** untrack requests on timeouts. It doesn't even have a notion of untracking requests itself. This is **important**, as it creates a deterministic workflow:

- Caller makes network request; if successful, gets an `eth.Request` instance
- Caller waits until it doesn't want to any more:
  - A. Response arrived, request is considered done.
  - B. Caller doesn't want to wait (e.g. too slow vs other peers), request is considered cancelled.
  - C. Caller has to abort (e.g. download cycle failed or shutdown), request is considered cancelled.
- Caller notifies the dispatcher to untrack the request.

This workflow is very easy to reason about as response delivery has it's dedicated flow/direction (dispatcher to caller via sink channel) and cleanup handling has it's own dedicated flow/direction (caller to dispatcher via method calls (internally channels)). There's no dirtying of flows by sending lifecycle events on the data channel.

### API

Although the dispatcher is part of an `eth.Peer`, it is not publicly callable. Since the `eth.Request` types need to be assembled - which is half protocol dependent, half dispatcher dependent - the requests are created and dispatched by the `peer.RequestXYZ` methods. Similarly, the `eth.Responses` are assembled and dispatched internally in the handlers.

However, since requests do need to be tracked and potentially cancelled, the `eth.Request` is public and returned via the `peer.RequestXYZ` methods. Similarly, it's simpler to also deliver `eth.Response` as a public type to the caller (via the request sink) since it allows bundling up a lot of useful metadata. The exact content was derived by what was necessary to support the fetchers and downloaders.

One downside of working with public `eth.Request` and `eth.Response` types is that the downloader and fetcher got locked in to the `eth` protocol, whereas previously the same code was used for `les` too. This is a curious issue, because on the surface it might be considered a downside - fixable with introducing interfaces (currently the old code was duplicated into les). But thinking about it mid-term, it's not really a good idea to complicate the codebase back again to support multiple protocols:

- The fetcher becomes useless after The Merge, since block propagation gets removed, so even the `eth` one becomes dead code.
- `les` becomes completely different after The Merge, since the chain head is available via the consensus node; and there's no reason to download entire past chain segments (currently we need it to sync).

As such, this PR assumes that both the duplicated fetcher and downloader in `les` will be altogether deleted after the merge. There's very little point in hacking the entire codebase into something abstract that has a half a year death sentence anyway.

### Refactors

The above described dispatcher which delivers responses directly to the callsite has the potential to simplify everything, but it does entail a significant refactor since everything was built on the bubble-through approach (`les` even does some accounting on top :scream:).

The PR does a full and clean refactor of the `downloader` to support the new model. Since `fast sync` is deprecated, that part of the code was deleted altogether (since refactoring it is painful), completely relying on `snap sync` from now on. Snap sync was not previously tested within the context of the downloader, so the PR also completely overhauls the downloader tests to use the real `core.BlockChain` code and `eth.Handlers` code as we can't meaningfully mock out the snapshot tree handlig. This is a huge change in the downloader, but it cleans *everything* up in preparation for the reverse syncer work for The Merge.

The PR did a murky-er refactor within the fetcher. Instead of rewriting the whole thing - at significant time cost - I've just did the bare minimum to hack in the new model. The thinking is that the fetcher is going away anyway after the merge, so don't waste time on fixing stuff on the death row.

The PR also cleanly integrated the dispatcher model into the `eth` header challenges.

### CODE REVIEW

File: eth/downloader/downloader.go
Comment (holiman) [inline eth/downloader/downloader.go:657]:
```suggestion
	// validated head of the chain. Check the pivot number and return,
```

File: eth/downloader/fetchers.go
Comment (holiman) [inline eth/downloader/fetchers.go:79]: If I'm reading it right, this entire method is identical to the method above, aside from this line. Could you make it like this: 

``` 
func (d *Downloader) fetchHeadersByHash(p *peerConnection, hash common.Hash, amount int, skip int, reverse bool) ([]*types.Header, error) {
	// Create the response sink and send the network request
	resCh := make(chan *eth.Response)
	req, err := p.peer.RequestHeadersByHash(hash, amount, skip, reverse, resCh)
        return handleHeaderRequest(req, err, resCh)
}
func (d *Downloader) fetchHeadersByNumber(p *peerConnection, number uint64, amount int, skip int, reverse bool) ([]*types.Header, error) {
	// Create the response sink and send the network request
	resCh := make(chan *eth.Response)
	req, err := p.peer.RequestHeadersByNumber(number, amount, skip, reverse, resCh)
        return handleHeaderRequest(req, err, resCh)
}
func (d *Downloader) handleHeaderRequest(p *peerConnection, req, err, resCh) ([]*types.Header, error) {
	start := time.Now()
        ...
}
```

File: eth/handler.go
Comment (holiman) [inline eth/handler.go:384]: Why not drop ? We asked for a header and got multiple. With request id's, there's no excuse for this.

Comment (karalabe) [reply eth/handler.go:384]: If you return an error as a result for a response processing it will result in a peer drop

Comment (karalabe) [reply eth/handler.go:384]: The downloader is doing manual peer dropping so that we don't hold up the receiver goroutine while the downloader hashes everything and verifies if the response matches the request (which allows for example headers, receipts and bodies to be processed concurrently).

Comment (karalabe) [reply eth/handler.go:384]: But here there's no expensive check, so we can insta verify and return an error if we dedct a protocol violation (i.e. not enough answers).

**Review by Conanza00 — APPROVED** [review]

**Comment by holiman** [comment]
With this change, I think we should make the request tracker more strict -- so we can rely on it for incoming messages. If we do that, we can do many checks earlier in the pipeline, e.g. checking block body before delivering upstream.

File: eth/downloader/fetchers.go
Comment (karalabe) [reply eth/downloader/fetchers.go:79]: I've made some additional changes in the reverse downloader PR on top of this. I'll make the modification there (pls don't forget) so I won't blow up the merge conflicts too much

File: core/blockchain.go
Comment (MariusVanDerWijden) [inline core/blockchain.go:760]: This looks like a rebasing artifact, had the same thing in my merge pr

Comment (karalabe) [reply core/blockchain.go:760]: What do you mean? This was deliberate in your PR too. Without this, the fast block is not updated on chain import at the end of the fast sync phase.

File: les/downloader/statesync.go
Comment (holiman) [inline les/downloader/statesync.go:613]: Either the clause can be removed entirely or this should be uncommented again?

Comment (karalabe) [reply les/downloader/statesync.go:613]: There's no fast sync in `les`, so either way it doesn't matter. I think I just needed to remove references to the main `downloader.SyncProcess` types, which have no notion of fast sync any more. I could have rewritten `les`'s downloader too, but figured that it's gonna happen anyway from scratch so better just keep the changes minimal and don't try to make actual "real" changes. I guess I left as a comment to keep te original code intact if we want to see what was there.

**Comment by holiman** [comment]
I ran the downloader-tests with `-race` -- I suspect this test is a bit timing-sensitive? 
```
--- PASS: TestMultiSynchronisation66Full (8.37s)
panic: failed to synchronise blocks: peer is stalling

goroutine 280419 [running]:
github.com/ethereum/go-ethereum/eth/downloader.testForkedSyncProgress.func3()
        /home/user/go/src/github.com/ethereum/go-ethereum/eth/downloader/downloader_test.go:1112 +0x17a
created by github.com/ethereum/go-ethereum/eth/downloader.testForkedSyncProgress
        /home/user/go/src/github.com/ethereum/go-ethereum/eth/downloader/downloader_test.go:1109 +0x7e5
FAIL    github.com/ethereum/go-ethereum/eth/downloader  246.671s
```

File: eth/downloader/downloader_test.go
Comment (karalabe) [reply eth/downloader/downloader_test.go:508]: Yes, this is unfortunately caused by all the tests running in parallel and too many generating chains. The downloader has built in timeouts that dynamically adjust to peers. If the downloader starts at one speed, and then another test spins up and hammers the CPU, the first one will be starved (compared to previously). Starvation will lead to the serving peer timing out and failing the test due to "stalling".

I can try to add a pre-generation step similar to what Felix did instead of the current lazy one. That would hopefully solve it, it's just wonky.

**Comment by karalabe** [comment]
> I ran the downloader-tests with `-race` -- I suspect this test is a bit timing-sensitive?
> 
> ```
> --- PASS: TestMultiSynchronisation66Full (8.37s)
> panic: failed to synchronise blocks: peer is stalling
> 
> goroutine 280419 [running]:
> github.com/ethereum/go-ethereum/eth/downloader.testForkedSyncProgress.func3()
>         /home/user/go/src/github.com/ethereum/go-ethereum/eth/downloader/downloader_test.go:1112 +0x17a
> created by github.com/ethereum/go-ethereum/eth/downloader.testForkedSyncProgress
>         /home/user/go/src/github.com/ethereum/go-ethereum/eth/downloader/downloader_test.go:1109 +0x7e5
> FAIL    github.com/ethereum/go-ethereum/eth/downloader  246.671s
> ```

Please see https://github.com/ethereum/go-ethereum/pull/23576#discussion_r744094239

If you enable the race detector, everything gets significantly slower, so yeah, kind of to be expected. The "solution" could be to abstract out the notion of `time` from the downloader and use a fake clock to drive it. We have 2-3 packages doing it, it's just a large rework and you'd need to mock out time in every submodule that the downloader depends on.

File: eth/fetcher/block_fetcher.go
Comment (MariusVanDerWijden) [inline eth/fetcher/block_fetcher.go:471]: This function now starts peer * hashes go routines instead of peer routines. I think that's by design, just wanted to make sure that this won't become a problem

File: eth/handler_eth.go
Comment (MariusVanDerWijden) [inline eth/handler_eth.go:71]: Here's still a todo

**Review by MariusVanDerWijden — COMMENTED** [review]
Tried my best to review it, I didn't properly check the `blockchain_test.go` and I could not super meaningfully review `fetchers_concurrent.go` The rest looks good to me. We should create some fuzzers for Snap packages btw

File: eth/fetcher/block_fetcher.go
Comment (karalabe) [reply eth/fetcher/block_fetcher.go:471]: Myeah, that's a good observation, alas delliberate.

Previously `fetchHeader` just sent a remote request and the reply came back from the outside from the handler on a separate goroutine.

With the request id style blocking APIs, I need something to wait for the reply to actually arrive. A better way would be to make a demultiplexer that listens on a sink channel and keeps consuming inbound responses. It requires a bit of a heavier internal rewrite that I didn't want to focus on yet. Though the question that's valid is whether that part should be done before a merge.

Comment (karalabe) [reply eth/fetcher/block_fetcher.go:471]: In practical terms, a new goroutine per request should be ok-ish, as we're only launching on request per missing data, so the only way to blow up the goroutine count is to blow up the block announcements; but that's limited specifically to avoid blowing up memory usage with spam. So we're still kind of bound by the old DoS protections, just it's +1 goroutine load.

File: eth/downloader/downloader_test.go
Comment (karalabe) [reply eth/downloader/downloader_test.go:496]: I made some changes to pregenerate the datasets and run the tests sequentially. Also reduced the dataset sizes a bit. Hopefully will stop dropping dead.

File: eth/handler_eth.go
Comment (karalabe) [reply eth/handler_eth.go:71]: Yes, I'll delete this after a few benchmark syncs. Just want to make sue these don't end up in live code.

File: eth/protocols/eth/dispatcher.go
Comment (karalabe) [reply eth/protocols/eth/dispatcher.go:131]: This was correct before. "wait until it is handled"

File: cmd/utils/flags.go
Comment (holiman) [inline cmd/utils/flags.go:208]: Might be that this change causes a lot of existing setups to refuse to start. Maybe we shoud deprecate `fast` and just internally translat it to `snap` for a while?

File: eth/protocols/eth/dispatcher.go
Comment (MariusVanDerWijden) [reply eth/protocols/eth/dispatcher.go:131]: Damn, the german dictionary site said something different than the english one

File: eth/downloader/downloader.go
Comment (holiman) [inline eth/downloader/downloader.go:985]: We have these various spread-out checks "did X deliver what we asked"? 
Right now, the dispatcher guarantees that 
- A response is matched with a request, 
- The response is of the correct type, and can be cast:ed to the expected response type. 

However, the early checks could do more: 
- For headers, it could verify the amount, numbers/hashes against the request, 
- For bodies, it could verify the sha of the body

Feels like a lot of the clauses here would go away if we did that, and maybe we could move some processing off to be multithreaded.

File: eth/downloader/downloader.go
Comment (MariusVanDerWijden) [inline eth/downloader/downloader.go:1122]: This would print both failed and succeeded

Comment (karalabe) [reply eth/downloader/downloader.go:1122]: Ugh, just added that while debugging :)) Lemme fix it.

File: eth/protocols/eth/dispatcher.go
Comment (holiman) [inline eth/protocols/eth/dispatcher.go:134]:
```suggestion
		// of picking one channel or the other. To avoid such cases delivering
```

File: eth/protocols/eth/dispatcher.go
Comment (karalabe) [inline eth/protocols/eth/dispatcher.go:179]: Oops, only do if it's not nil

File: eth/downloader/fetchers_concurrent.go
Comment (holiman) [inline eth/downloader/fetchers_concurrent.go:284]: With this change here, afaict the comment is now misleading. We don't cancel it so it's not considered in-flight anymore, since we no longer call `req.Close()` on it. I suppose the caller keeps waiting for the reply, even though this internally has been moved to stale? And after a little while longer, the peer will be rejected, prompting the stale request to be `Close`d and rescheduled?

File: eth/protocols/eth/dispatcher.go
Comment (holiman) [inline eth/protocols/eth/dispatcher.go:200]: s/pendign/pending

File: eth/protocols/eth/dispatcher.go
Comment (holiman) [inline eth/protocols/eth/dispatcher.go:214]: What happens if the response to a `stale` request comes in -- we just drop it, right? I don't quite follow the logic. So we shuffle it over from `pending` to `stale`, but we still don't (immediately) reschedule it, but neither do we handle a late response? 
I must have missed something, but it seems like moving something to `stale` unconditionally means it _will_ time out a bit later, leading to the peer being rejected.

Comment (holiman) [reply eth/protocols/eth/dispatcher.go:214]: Ah, my bad. The `pending` here is dispatcher-local pending, it has no relation to the pending/stales in the downloader.

**Review by holiman — APPROVED** [review]
LGTM

**Pull request merged**
